go/pkg/contextmd/contextmd.go

```diff
@@ -177,33 +177,33 @@ func mergeSet(set map[string]struct{}) string {
 
 // capToLimit ensures total length does not exceed max header size.
 func capToLimit(m *Metadata, limit int) *Metadata {
-	total := len(m.ToolName) + len(m.ToolVersion) + len(m.ActionID) + len(m.InvocationID) + len(m.CorrelatedInvocationID)
+	total := len(m.ToolName) + len(m.ToolVersion) + len(m.ActionID) + len(m.InvocationID) + len(m.CorrelatedInvocationsID)
 	excess := total - limit
 	if excess <= 0 {
 		return m
 	}
-	// We ignore the tool name, because in practice this is a
-	// very short constant which makes no sense to truncate.
-	diff := len(m.ActionID) - len(m.InvocationID)
-	if diff > 0 {
-		if diff > excess {
-			m.ActionID = m.ActionID[:len(m.ActionID)-excess]
-		} else {
-			m.ActionID = m.ActionID[:len(m.ActionID)-diff]
-			rem := (excess - diff + 1) / 2
-			m.ActionID = m.ActionID[:len(m.ActionID)-rem]
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-rem]
-		}
-	} else {
-		diff = -diff
-		if diff > excess {
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-excess]
-		} else {
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-diff]
-			rem := (excess - diff + 1) / 2
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-rem]
-			m.ActionID = m.ActionID[:len(m.ActionID)-rem]
+	truncateStrings(limit, &m.ActionID, &m.InvocationID, &m.CorrelatedInvocationsID, &m.ToolName, &m.ToolVersion)
+	return m
+}
+
+// truncateStrings remove one char from the longest str at a time until the
+// total length of all the strings is less than limit value.
+func truncateStrings(limit int, inputs ...*string) {
+	total := 0
+	for _, s := range inputs {
+		total += len(*s)
+	}
+	if total <= limit {
+		return
+	}
+	for total > limit {
```

Comment:
This loop might end up being expensive considering it's executed at least once for every action. I think it's safer to limit this PR to what's needed to satisfy its title and leave this refactor to another one.